packages/gestalt/src/Tooltip.js

```diff
@@ -3,62 +3,128 @@
 import * as React from 'react';
 import Controller from './Controller.js';
 import Text from './Text.js';
+import Link from './Link.js';
 import Box from './Box.js';
 
 const noop = () => {};
 
 type Props = {|
   children: React.Node,
+  href?: string,
+  hrefText?: string,
   inline?: boolean,
   text: string,
 |};
 
-type State = {|
-  hovered: boolean,
-|};
+const initialState = { hoveredIcon: false, hoveredText: false, isOpen: false };
+
+const reducer = (state, action) => {
+  switch (action.type) {
+    case 'hoverInIcon':
+      return {
+        ...state,
+        hoveredIcon: true,
+        isOpen: true,
+      };
+    case 'hoverInText':
+      return {
+        ...state,
+        hoveredText: true,
+        isOpen: true,
+      };
+    case 'hoverOutIcon':
+      return {
+        ...state,
+        hoveredIcon: false,
+        ...(!state.hoveredText ? { isOpen: false } : {}),
+      };
+    case 'hoverOutText':
+      return {
+        ...state,
+        hoveredText: false,
+        ...(!state.hoveredIcon ? { isOpen: false } : {}),
+      };
+    default:
+      throw new Error();
+  }
+};
 
-export default class Tooltip extends React.Component<Props, State> {
-  state = {
-    hovered: false,
+export default function Tooltip({ children, href, hrefText, inline, text }) {
+  const [state, dispatch] = React.useReducer(reducer, initialState);
+  const { isOpen } = state;
+
+  const childRef = React.useRef<?HTMLDivElement>(null);
+  const { current: anchor } = childRef;
+
+  const handleIconMouseEnter = () => {
+    dispatch({ type: 'hoverInIcon' });
   };
 
-  childRef: {| current: null | React.ElementRef<'div'> |} = React.createRef();
+  const handleIconMouseLeave = () => {
+    setTimeout(() => {
+      dispatch({ type: 'hoverOutIcon' });
+    }, 75);
+  };
 
-  handleMouseEnter = () => this.setState({ hovered: true });
+  const handleTextMouseEnter = () => {
+    dispatch({ type: 'hoverInText' });
+  };
 
-  handleMouseLeave = () => this.setState({ hovered: false });
+  const handleTextMouseLeave = () => {
+    setTimeout(() => {
+      dispatch({ type: 'hoverOutText' });
+    }, 75);
+  };
 
-  render() {
-    const { children, inline, text } = this.props;
-    const { hovered } = this.state;
-    const { current: anchor } = this.childRef;
+  const allowTooltipHover = href && hrefText;
 
-    return (
-      <Box display={inline ? 'inlineBlock' : 'block'}>
-        <Box
-          onMouseEnter={this.handleMouseEnter}
-          onMouseLeave={this.handleMouseLeave}
-          ref={this.childRef}
+  return (
+    <Box display={inline ? 'inlineBlock' : 'block'}>
+      <Box
+        ref={childRef}
+        onMouseEnter={handleIconMouseEnter}
+        onMouseLeave={handleIconMouseLeave}
+        onFocus={handleIconMouseEnter}
+        onBlur={handleIconMouseLeave}
+        role="tooltip-container"
```

Comment:
@anniejiwon this is an invalid role, can we please remove it? https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques#Roles

We also have an `onFocus` and `onBlur` event but no `tabIndex` set to `0` - could we add that `tabIndex`?